### PR TITLE
Create a Microsoft.XmlSerializer.Generator Project

### PR DESCRIPTION
**Pull request by huanwu**
Create a Microsoft.XmlSerializer.Generator as a base project for SGen tool to be implemented. The project share the code with System.Private.Xml. And also move some codegen only code from system.private.xml to the new project.

#17579 

@weshaggard @zhenlan @mconnew @shmao

### CODE REVIEW

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj [hunk 1]
```diff
@@ -0,0 +1,90 @@
+﻿<Project Sdk="Microsoft.NET.Sdk">
```
Comment (weshaggard) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj]: This looks like it is trying to use the new CLI tools which currently corefx doesn't support using.

**Comment by weshaggard** [comment]
We should try and add some test code for this.

At some point you will also want to add a pkgproj in order to start producing a package for it.

**Comment by zhenlan** [comment]
> using System;

Please add header (copyright etc.) to this file.

Please also change the file name to match the class name, in this case, Sgen.cs (capital 'S').

---
Refers to: src/Microsoft.XmlSerializer.Generator/src/sgen.cs:1 in e1fbd07. [](commit_id = e1fbd07a1c7db9a54391bc7bee1bced4e9ffb0de, deletion_comment = False)

**Comment by zhenlan** [comment]
> }

I cannot believe anyone is actually using this. We don't support AppSetting on .NET Core. It's okay you clean this up in a separate PR if that makes it easier for you.

---
Refers to: src/System.Private.Xml/src/System/Xml/Serialization/AppSettings.cs:29 in e1fbd07. [](commit_id = e1fbd07a1c7db9a54391bc7bee1bced4e9ffb0de, deletion_comment = False)

File: src/System.Private.Xml/src/System/Xml/Serialization/CodeGenerationoptions.cs [hunk 2]
```diff
@@ -2,7 +2,13 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#if XMLSERIALIZERGENERATOR
+using System;
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/CodeGenerationoptions.cs]: I suppose this `using` is not really needed?

File: src/System.Private.Xml/src/System/Xml/Serialization/Globals.cs [hunk 3]
```diff
@@ -7,7 +7,13 @@
 using System.Reflection;
 
 
+#if XMLSERIALIZERGENERATOR
+using System;
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/Globals.cs]: Why this `using` is only needed for SGEN?

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs [hunk 4]
```diff
@@ -509,6 +515,7 @@ private SpecialMapping ImportSpecialMapping(Type type, TypeDesc typeDesc, string
                     XmlQualifiedName qname = serializableMapping.XsiType;
                     if (qname != null && !qname.IsEmpty)
                     {
+                        #if !XMLSERIALIZERGENERATOR
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlReflectionImporter.cs]: Please fix the indentation.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj (hunk 1 above)
Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj]: @huanwu will change this project to be like System.CodeDom.

**Comment by zhenlan** [comment]
> We should try and add some test code for this.

Yes. Because of the large number of files involved, this is meant to be a baseline check-in to make future incremental changes easier be reviewed. In future check-ins, we will add tests.

> At some point you will also want to add a pkgproj in order to start producing a package for it.

@huanwu can you see if you can add a `pkg` folder as part of this PR. @weshaggard we will not need a `ref` folder, correct?

**Comment by weshaggard** [comment]
> @weshaggard we will not need a ref folder, correct?

You shouldn't need a ref for this.

**Comment by shmao** [comment]
We probably don't want to have `Microsoft.XmlSerializer.Generator/src/Resources/SR.Designer.cs`. By having this file, we would need to maintain two resources for `System.Private.Xml` in the future. We should reuse the resource used by `System.Private.Xml`.

File: src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs [hunk 5]
```diff
@@ -164,7 +175,11 @@ public static unsafe byte[] Decode(char[] chars, bool allowOddChars)
 
             if (hasHalfByteCached && !allowOddChars)
             {
+#if XMLSERIALIZERGENERATOR
+                throw new XmlException(SR.Format(SR.Xml_InvalidBinHexValueOddCount, new string(chars)));
+#else
                 throw new XmlException(SR.Xml_InvalidBinHexValueOddCount, new string(chars));
+#endif
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: We shouldn't change this.

File: src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs [hunk 6]
```diff
@@ -218,10 +233,14 @@ public static unsafe byte[] Decode(char[] chars, bool allowOddChars)
                 }
                 else
                 {
+#if XMLSERIALIZERGENERATOR
+                    throw new XmlException(SR.Format(SR.Xml_InvalidBinHexValueOddCount, new string(pChars, 0, (int)(pCharsEndPos - pChars))));
+#else
                     throw new XmlException(SR.Xml_InvalidBinHexValue, new string(pChars, 0, (int)(pCharsEndPos - pChars)));
+#endif
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: We shouldn't change this.

File: src/System.Private.Xml/src/System/Xml/Serialization/CodeGenerationoptions.cs (hunk 2 above)
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/CodeGenerationoptions.cs]: For XMLSERIALIZERGENERATOR, we need `using System` since the new namespace, `Microsoft.XmlSerializer.Generator`, is not under `System.XXX`. But I think we can put the `using` outside `#if`.

File: src/System.Private.Xml/src/System/Xml/Serialization/IXmlSerializable.cs [hunk 7]
```diff
@@ -2,7 +2,13 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#if XMLSERIALIZERGENERATOR
+using System.Xml;
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/IXmlSerializable.cs]: Let's put this outside `#if XMLSERIALIZERGENERATOR` to reduce the code being forked.

File: src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs (hunk 5 above)
Comment (huanwu) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: I have to make a copy of strings.resx since the current build system always auto add the strings.resx under the resource folder of the project to the project. That means you are not allowed to refer a strings.resx in your project file.

Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: We can override the path to the .resx file by setting StringResourcesPath in csproj. See https://github.com/dotnet/corefx/blob/master/Tools-Override/Build.Common.targets#L177 for details.

File: src/Microsoft.XmlSerializer.Generator/src/Resources/SR.Designer.cs [hunk 8]
```diff
@@ -0,0 +1,9858 @@
+﻿//------------------------------------------------------------------------------
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/src/Resources/SR.Designer.cs]: Do we need this file? If not, let's remove it.

File: src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs [hunk 9]
```diff
@@ -218,10 +233,14 @@ public static unsafe byte[] Decode(char[] chars, bool allowOddChars)
                 }
                 else
                 {
+#if XMLSERIALIZERGENERATOR
+                    throw new XmlException(SR.Format(SR.Xml_InvalidBinHexValue, new string(pChars, 0, (int)(pCharsEndPos - pChars))));
+#else
                     throw new XmlException(SR.Xml_InvalidBinHexValue, new string(pChars, 0, (int)(pCharsEndPos - pChars)));
+#endif
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: We don't need this change.

File: src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs (hunk 5 above)
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: We don't need this change.

File: src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs [hunk 10]
```diff
@@ -89,8 +95,8 @@ internal TempAssembly(XmlMapping[] xmlMappings, Type[] types, string defaultName
 #endif
 
 #if DEBUG
-            // use exception in the place of Debug.Assert to avoid throwing asserts from a server process such as aspnet_ewp.exe
-            if (_assembly == null) throw new InvalidOperationException(SR.Format(SR.XmlInternalErrorDetails, "Failed to generate XmlSerializer assembly, but did not throw"));
+                // use exception in the place of Debug.Assert to avoid throwing asserts from a server process such as aspnet_ewp.exe
+                if (_assembly == null) throw new InvalidOperationException(SR.Format(SR.XmlInternalErrorDetails, "Failed to generate XmlSerializer assembly, but did not throw"));
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/Compilation.cs]: The indention seems incorrect.

File: src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 11]
```diff
@@ -14,8 +14,15 @@
 using System.Xml.Extensions;
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs]: We shouldn't need `ReflectionXmlSerializationReader` or `ReflectionXmlSerializationWriter`. Can you please open an issue for tracking excluding these files from Sgen. We would also need to address the conditions for using reflection based serialization.

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlCountingReader.cs [hunk 12]
```diff
@@ -2,13 +2,20 @@
...
+#else
 namespace System.Xml.Serialization
+#endif
 {
     using System.IO;
     using System.Collections;
     using System.Text;
     using System;
     using System.Xml.Schema;
+#if XMLSERIALIZERGENERATOR
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlCountingReader.cs]: Remove the `#if`.

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriterILGen.cs [hunk 13]
```diff
@@ -14,7 +14,13 @@
...
 
 #if !uapaot
+#if XMLSERIALIZERGENERATOR
+using System.Xml;
+
+namespace Microsoft.XmlSerializer.Generator
+#else
 namespace System.Xml.Serialization
+#endif
 {
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializationWriterILGen.cs]: This file is not included in Sgen anymore. Can you please revert the changes in this file?

The same for ReaderILGen, ILGen ..

File: src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializerVersionAttribute.cs [hunk 14]
```diff
@@ -2,7 +2,11 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#if XMLSERIALIZERGENERATOR
+namespace Microsoft.XmlSerializer.Generator
+#else
 namespace System.Xml.Serialization
+#endif
 {
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/XmlSerializerVersionAttribute.cs]: Does Sgen need to include this file? As I recalled, you changed the code to use System.Xml.Serialization.XmlSerializerVersionAttribute explicitly. I think we may not need this.

The same for `XmlSerializerAssemblyAttribute`.

File: src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs [hunk 15]
```diff
@@ -218,10 +233,14 @@ public static unsafe byte[] Decode(char[] chars, bool allowOddChars)
...
                 }
 
-                if (hasHalfByteCached)
+                    if (hasHalfByteCached)
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/BinHexDecoder.cs]: Please fix the indention.

File: src/System.Private.Xml/src/System/Xml/Serialization/Globals.cs [hunk 16]
```diff
@@ -7,8 +7,14 @@
 using System.Reflection;
 
 
+#if XMLSERIALIZERGENERATOR
+namespace Microsoft.XmlSerializer.Generator
+#else
 namespace System.Xml.Serialization
+#endif
 {
+    using System;
```
Comment (shmao) [inline src/System.Private.Xml/src/System/Xml/Serialization/Globals.cs]: Nit: Can you please put this along with other `using` directives? Thanks.

File: src/Microsoft.XmlSerializer.Generator/dir.props [hunk 17]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\dir.props" />
+  <PropertyGroup>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
+    <IsNETCoreApp>true</IsNETCoreApp>
+    <IsNETCoreAppRef>false</IsNETCoreAppRef>
+    <IsUAP>true</IsUAP>
```
Comment (shmao) [inline src/Microsoft.XmlSerializer.Generator/dir.props]: Shoud <IsUAP> be false? @zhenlan

Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/dir.props]: Yes, I think so (IsUAP=false).

---
In reply to: [109057070](https://github.com/dotnet/corefx/pull/17615#discussion_r109057070) [](ancestors = 109057070)

File: src/Microsoft.XmlSerializer.Generator/dir.props [hunk 18]
```diff
@@ -0,0 +1,11 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\dir.props" />
+  <PropertyGroup>
+    <AssemblyVersion>4.0.0.0</AssemblyVersion>
```
Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/dir.props]: Let's start AssemblyVersion with 1.0.0.0.

File: src/Microsoft.XmlSerializer.Generator/src/Configurations.props [hunk 19]
```diff
@@ -0,0 +1,9 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <PropertyGroup>
+    <BuildConfigurations>
+      netfx;
```
Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/src/Configurations.props]: We probably don't need to build for netfx.

File: src/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj [hunk 20]
```diff
@@ -0,0 +1,244 @@
...
+    </Compile>
+    <Compile Include="$(SourceDir)\System\Xml\XmlReservedNs.cs">
+      <Link>System\Xml\XmlReservedNs.cs</Link>
+    </Compile>
+  </ItemGroup>
+  <ItemGroup>
+    <Compile Include="Microsoft\XmlSerializer\Generator\XmlSerializationGeneratedCode.cs" />
+    <Compile Include="Microsoft\XmlSerializer\Generator\XmlSerializationReader.cs" />
+    <Compile Include="Microsoft\XmlSerializer\Generator\XmlSerializationWriter.cs" />
+    <Compile Include="sgen.cs" />
```
Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/src/Microsoft.XmlSerializer.Generator.csproj]: Change sgen.cs to Sgen.cs to match class name case.

**Comment by zhenlan** [comment]
> 

NIT: remove empty lines.

---
Refers to: src/Microsoft.XmlSerializer.Generator/tests/SGenTests.cs:23 in 9414468. [](commit_id = 9414468445a1902ffb4651f20616e84a172774a2, deletion_comment = False)

File: src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj [hunk 21]
```diff
@@ -0,0 +1,17 @@
...
+    <ProjectGuid>{0D1E2954-A5C7-4B8C-932A-31EB4A96A726}</ProjectGuid>
+  </PropertyGroup>
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netstandard-Debug|AnyCPU'" />
+  <PropertyGroup Condition="'$(Configuration)|$(Platform)'=='netstandard-Release|AnyCPU'" />
+  <ItemGroup>
+  <Compile Include=".\SGenTests.cs" />
+  </ItemGroup>
+<ItemGroup>
+
+    <ProjectReference Include="..\src\Microsoft.XmlSerializer.Generator.csproj" />
```
Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj]: NIT: the indentation and empty line.

File: src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj [hunk 22]
```diff
@@ -0,0 +1,17 @@
+<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" xmlns="http://schemas.microsoft.com/developer/msbuild/2003" DefaultTargets="Build">
+  <Import Project="$([MSBuild]::GetDirectoryNameOfFileAbove($(MSBuildThisFileDirectory), dir.props))\dir.props" />
+  <PropertyGroup>
+    <ProjectGuid>{0D1E2954-A5C7-4B8C-932A-31EB4A96A726}</ProjectGuid>
```
Comment (zhenlan) [inline src/Microsoft.XmlSerializer.Generator/tests/Microsoft.XmlSerializer.Generator.Tests.csproj]: Please create a new Guild. Don't copy this from CodeDom.

File: src/System.Private.Xml/src/System/Xml/Serialization/NameTable.cs [hunk 23]
```diff
@@ -2,7 +2,14 @@
 // The .NET Foundation licenses this file to you under the MIT license.
 // See the LICENSE file in the project root for more information.
 
+#if XMLSERIALIZERGENERATOR
+using System.Xml;
+using System;
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/NameTable.cs]: Put all using together.

File: src/System.Private.Xml/src/System/Xml/Serialization/PrimitiveXmlSerializers.cs [hunk 24]
```diff
@@ -2,10 +2,21 @@
...
 {
+    using System;
+
+#if XMLSERIALIZERGENERATOR
+    internal class XmlSerializationPrimitiveWriter : Microsoft.XmlSerializer.Generator.XmlSerializationWriter
+#else
     internal class XmlSerializationPrimitiveWriter : System.Xml.Serialization.XmlSerializationWriter
+#endif
     {
+
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/PrimitiveXmlSerializers.cs]: We probably can combine the two #if sections into one and put the using outside.

File: src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs [hunk 25]
```diff
@@ -14,8 +14,15 @@
...
 using System.Xml.Schema;
 
+#if XMLSERIALIZERGENERATOR
+namespace Microsoft.XmlSerializer.Generator
+#else
 namespace System.Xml.Serialization
+#endif
 {
+    using System.Xml;
+    using System.Xml.Serialization;
```
Comment (zhenlan) [inline src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs]: I will stop comment on this :)  Pease put all using together in this file and others.

**Comment by huanwu** [comment]
@dotnet-bot test Windows_NT Debug AllConfigurations Build

File: src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs (hunk 11 above)
Comment (huanwu) [inline src/System.Private.Xml/src/System/Xml/Serialization/ReflectionXmlSerializationReader.cs]: Opened issue #17752

File: src/Microsoft.XmlSerializer.Generator/dir.props [hunk 26]
```diff
@@ -0,0 +1,13 @@
+﻿<?xml version="1.0" encoding="utf-8"?>
+<Project ToolsVersion="14.0" DefaultTargets="Build" xmlns="http://schemas.microsoft.com/developer/msbuild/2003">
+  <Import Project="..\dir.props" />
+  <PropertyGroup>
+    <PackageVersion>1.0.0</PackageVersion>
+    <SkipValidatePackage>true</SkipValidatePackage>
+    <AssemblyVersion>1.0.0.0</AssemblyVersion>
+    <IsNETCoreApp>true</IsNETCoreApp>
```
Comment (weshaggard) [inline src/Microsoft.XmlSerializer.Generator/dir.props]: This should not set IsNETCoreApp=true. In fact you should just remove all these Is* properties from this project.
